packages/apiserver/src/resources/item/item.service.ts

```diff
@@ -2,77 +2,99 @@ import { Injectable } from '@nestjs/common';
-      });
-      if (status.status === StatusTypes.PROLONG && type !== StatusTypes.FREE) {
-        throw new ErrorBook('This book has been extended');
+
+      if (material.currentStatus.status !== StatusTypes.FREE) {
+        throw new ErrorBook(`This book can't be claimed. Ask the manager!`);
+      }
+
+      return this.insertStatus(
+        {
+          status: StatusTypes.BUSY,
+          material_id: material.id,
+          person_id,
+          returnDate: () => `NOW() + INTERVAL '${this.periodOfKeeping} days'`,
+        },
+        manager
+      );
+    });
+  }
+
+  async putOnShelf(returnBookInput: BookInput): Promise<Status> {
+    return this.connection.transaction(async (manager) => {
+      const { identifier, person_id } = returnBookInput;
+      const material = await this.getMaterialByIdentifier(identifier, manager);
+
+      if (!material) {
+        throw new ErrorBook('This item is not registered in the library');
       }
-      if (status.status === type) {
+
+      if (
+        material.currentStatus.status !== StatusTypes.BUSY &&
+        material.currentStatus.status !== StatusTypes.PROLONG &&
+        material.currentStatus.status !== StatusTypes.PENDING
+      ) {
         throw new ErrorBook(
-          `This book is ${type.toLocaleLowerCase()}. Ask the manager!`
+          `This book can't be put on shelf. Ask the manager!`
         );
       }
-      const newStatusObj = await statusRepository.create({
-        status: type,
-        material_id: status.material_id,
-        person_id,
-      });
-      const newStatus = await statusRepository.save(newStatusObj);
-      await queryRunner.commitTransaction();
-      return newStatus;
-    } catch (e) {
-      await queryRunner.rollbackTransaction();
-      return {
-        message: e.message,
-      };
-    } finally {
-      await queryRunner.release();
-    }
-  }
 
-  async claim(claimBookInput: BookInput) {
-    return this.claimOrReturnOperations(claimBookInput, StatusTypes.BUSY);
+      return this.insertStatus(
+        {
+          status: StatusTypes.FREE,
+          material_id: material.id,
+          person_id,
+        },
+        manager
+      );
+    });
   }
 
-  async return(returnBookInput: BookInput) {
-    return this.claimOrReturnOperations(returnBookInput, StatusTypes.FREE);
-  }
-  async reject(returnBookInput: BookInput) {
-    return this.claimOrReturnOperations(returnBookInput, StatusTypes.REJECTED);
+  async reject(rejectBookInput: BookInput): Promise<Status> {
+    return this.connection.transaction(async (manager) => {
+      const { identifier, person_id } = rejectBookInput;
+      const material = await this.getMaterialByIdentifier(identifier, manager);
+
+      if (!material) {
+        throw new ErrorBook('This item is not registered in the library');
+      }
+
+      if (material.currentStatus.status !== StatusTypes.PENDING) {
+        throw new ErrorBook(`This book can't be rejected. Ask the manager!`);
+      }
```

Comment:
Only manager can reject the book currently, but minor

Reply:
Fixed